### PR TITLE
[AssetMapper] Load es-module-shims only if importmap is not supported

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

Let's not load the polyfill if the browser already supports importmaps

### CODE REVIEW

File: src/Symfony/Component/AssetMapper/Tests/ImportMap/ImportMapRendererTest.php [hunk 1]
```diff
@@ -132,7 +132,7 @@ public function testCustomScriptAttributes()
         ]);
         $html = $renderer->render([]);
         $this->assertStringContainsString('<script type="importmap" something data-turbo-track="reload">', $html);
-        $this->assertStringContainsString('<script async src="https://polyfillUrl.example" something data-turbo-track="reload"></script>', $html);
```
Comment (stof) [inline src/Symfony/Component/AssetMapper/Tests/ImportMap/ImportMapRendererTest.php:135]: you lost the assertion on `data-turbo-track="reload"`

Comment (nicolas-grekas) [reply src/Symfony/Component/AssetMapper/Tests/ImportMap/ImportMapRendererTest.php:135]: oops, fixed

**Comment by newQuery** [comment]
I'd suggest to create two separate methods `formatAttributes` & `generateAttributeScript` to simplify the code 

<details>

```php
if ($polyfillPath) {
    $polyfillAttributes = array_merge([
        'src' => $polyfillPath,
        'async' => true,
    ], $attributes, $this->scriptAttributes);

    // Add security attributes if using the default polyfill
    if ($polyfillPath === self::DEFAULT_ES_MODULE_SHIMS_POLYFILL_URL) {
        $polyfillAttributes = array_merge([
            'crossorigin' => 'anonymous',
            'integrity' => self::DEFAULT_ES_MODULE_SHIMS_POLYFILL_INTEGRITY,
        ], $polyfillAttributes);
    }

    $output .= '<script async' . $this->formatAttributes($scriptAttributes) . '>
        if (!HTMLScriptElement.supports("importmap")) {
            (function() {
                const script = document.createElement("script");
                ' . $this->generateAttributeScript($polyfillAttributes) . '
                document.head.appendChild(script);
            })();
        }
    </script>';
}
```

The methods: 

```php
private function formatAttributes(array $attributes): string
{
    $formatted = '';
    foreach ($attributes as $name => $value) {
        $formatted .= ' ' . htmlspecialchars($name, \ENT_NOQUOTES) . '="' . htmlspecialchars($value, \ENT_NOQUOTES) . '"';
    }
    return $formatted;
}

private function generateAttributeScript(array $attributes): string
{
    $script = '';
    foreach ($attributes as $name => $value) {
        $name = addslashes(htmlspecialchars($name, \ENT_NOQUOTES));
        $value = addslashes(htmlspecialchars($value, \ENT_NOQUOTES));
        $script .= "\n    script.setAttribute('{$name}', '{$value}');";
    }
    return $script;
}
```

</details>

Improve code readability, maintainability, and reusability

**Comment by newQuery** [comment]
I would also allow developers to configure whether es-module-shims should be loaded, providing more flexibility

**Comment by nicolas-grekas** [comment]
> I'd suggest to create two separate methods formatAttributes & generateAttributeScript to simplify the code

Factorized: I leveraged the already existing method createAttributeString

> I would also allow developers to configure whether es-module-shims should be loaded, providing more flexibility

Not sure this would provide any benefits. Options are complexity. Anyway, that's a discussion for another PR/ issue.

File: src/Symfony/Component/AssetMapper/ImportMap/ImportMapRenderer.php [hunk 2]
```diff
@@ -114,21 +114,26 @@ public function render(string|array $entryPoint, array $attributes = []): string
...
-                ] + $attributes);
+                ] + $polyfillAttributes;
             }
 
             $output .= <<<HTML
-
-                <!-- ES Module Shims: Import maps polyfill for modules browsers without import maps support -->
-                <script async src="$url"$polyfillAttributes></script>
+                <script async$scriptAttributes>
+                if (!HTMLScriptElement.supports || !HTMLScriptElement.supports('importmap')) (function () {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/AssetMapper/ImportMap/ImportMapRenderer.php:129]: I confirmed this works using Browerstack :shipit:

Comment (chalasr) [reply src/Symfony/Component/AssetMapper/ImportMap/ImportMapRenderer.php:129]: Soon in Nicolas' bio "GPT+BrowserStack powered JS dev" 😁

**Pull request merged**